src/Orleans/Serialization/SerializationManager.cs

```diff
@@ -110,64 +112,57 @@ public static object GetUninitializedObjectWithFormatterServices(Type type)
+        internal IntValueStatistic TotalTimeInSerializer;
+
+        internal CounterStatistic FallbackSerializations;
+        internal CounterStatistic FallbackDeserializations;
+        internal CounterStatistic FallbackCopies;
+        internal CounterStatistic FallbackSerTimeStatistic;
+        internal CounterStatistic FallbackDeserTimeStatistic;
+        internal CounterStatistic FallbackCopiesTimeStatistic;
 
         internal static int LARGE_OBJECT_LIMIT = Constants.LARGE_OBJECT_HEAP_THRESHOLD;
 
-        [ThreadStatic]
-        private static SerializationContext serializationContext;
-
-        private static SerializationContext CurrentSerializationContext
-            => serializationContext ?? (serializationContext = new SerializationContext(null));
+        private readonly ThreadLocal<SerializationContext> serializationContext;
         
-        [ThreadStatic]
-        private static DeserializationContext deserializationContext;
+        private readonly ThreadLocal<DeserializationContext> deserializationContext;
+        private readonly IServiceProvider serviceProvider;
+        private IRuntimeClient runtimeClient;
+
+        internal IRuntimeClient RuntimeClient
+            => this.runtimeClient ?? (this.runtimeClient = this.serviceProvider.GetService<IRuntimeClient>());
 
-        private static DeserializationContext CurrentDeserializationContext
-            => deserializationContext ?? (deserializationContext = new DeserializationContext(null));
+        internal IServiceProvider ServiceProvider => this.serviceProvider;
 
         #endregion
 
         #region Static initialization
-
-        public static void InitializeForTesting(List<TypeInfo> serializationProviders = null, TypeInfo fallbackType = null)
+        
+        public SerializationManager(IServiceProvider serviceProvider, IMessagingConfiguration config)
+            : this(serviceProvider, config.SerializationProviders, config.FallbackSerializationProvider)
         {
-            try
-            {
-                RegisterBuiltInSerializers();
-                BufferPool.InitGlobalBufferPool(new MessagingConfiguration(false));
-                RegisterSerializationProviders(serializationProviders);
-                fallbackSerializer = GetFallbackSerializer(fallbackType);
-            }
-            catch (ReflectionTypeLoadException ex)
-            {
-                throw ex.Flatten();
-            }
         }
 
-        internal static void Initialize(List<TypeInfo> serializationProviders, TypeInfo fallbackType = null)
+        internal SerializationManager(IServiceProvider serviceProvider, List<TypeInfo> serializationProviders = null, TypeInfo fallbackType = null)
         {
+            this.serializationContext = new ThreadLocal<SerializationContext>(() => new SerializationContext(this));
+            this.deserializationContext = new ThreadLocal<DeserializationContext>(() => new DeserializationContext(this));
```

Comment:
I'm not *too* worried that we are passing a reference to `this` in this constructor (although it could also be done in a factory method, but it's OK regardless), but could we at least make this assignment at the end of the constructor, so that we make sure nobody uses the field before the object is done constructing.